src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,573 @@
+#include "subquery_to_join_rule.hpp"
+
+#include <map>
+#include <memory>
+#include <set>
+
+#include "cost_model/cost_model_logical.hpp"
+#include "expression/abstract_expression.hpp"
+#include "expression/abstract_predicate_expression.hpp"
+#include "expression/binary_predicate_expression.hpp"
+#include "expression/expression_functional.hpp"
+#include "expression/expression_utils.hpp"
+#include "expression/in_expression.hpp"
+#include "expression/lqp_column_expression.hpp"
+#include "expression/lqp_subquery_expression.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/aggregate_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "statistics/table_statistics.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+namespace {
+
+/**
+ * Info about a predicate that needs to be pulled up into a join predicate.
+ */
+struct PredicateInfo {
```

Comment:
The idea was to have a struct where the operands are ordered by the subtrees they come from. Since we only use it in one vector in `PredicatePullUpInfo`, we can just have it as a comment there.

Reply:
On a related note: `BinaryPredicateExpression` does not seem to be very usable without a `shared_ptr`/`unique_ptr` around it. For me it was throwing compile errors every time the move constructor was used, for example when swapping or when assigning an `std::optional<BinaryPredicateExpression>`. Is this intentional?

Reply:
Not by design, but as the expression calsses have some const members they cannot be copied/moved into. If you need swapping/assigning you're probably actually better off using your own struct to represent the expression. The expression classes could be turned into copyable types (and probably should at some point) but I wouldn't want to do that as part of this PR if there is only a single use case for this.

Reply:
I wrapped them in shared_ptrs for now. Since `JoinNode` expect its predicates as `shared_ptr<AbstractExpression>` anyways, there shouldn't be much overhead except for a few additional indirections.